Refactor cancel button spam for Chrysler

xps-genesis opened the pull request:
Modify button spam logic to avoid acc faults when pcm_cancel_cmd = True

xps-genesis committed:
refactor cancel button spam for Chrysler

xps-genesis committed:
refactor cancel button spam for Chrysler

xps-genesis committed:
refactor cancel button spam for Chrysler

adeebshihadeh commented on selfdrive/car/chrysler/carcontroller.py:
```diff
@@ -45,10 +46,14 @@ def update(self, enabled, CS, actuators, pcm_cancel_cmd, hud_alert):
+      self.wheel_button_counter_prev = CS.wheel_button_counter
 
-    if pcm_cancel_cmd:
+    if pcm_cancel_cmd and (self.ccframe % 10 < 5) and wheel_button_counter_change:
```
what does `self.ccframe % 10 < 5` do?

xps-genesis replied:
sends cancel for 50msec and gives a 50msec break before resending for 50msec

adeebshihadeh replied:
how often is the first message not accepted?

xps-genesis replied:
don't have any data, but have had issue with continuous spamming.

adeebshihadeh replied:
can you try removing the that condition? I'd also like some information on how often the first cancel isn't taken.

xps-genesis replied:
alright.  u probably can gather that info from all the chrysler users. my vehicle is a my21 pacifica and it had more issues with than older pacificas.

j-vanetten replied:
It shouldn't be needed as the cancel button is accepted on the down press and not up.  We should continue to send the press until OP disengages.

xps-genesis replied:
thanks

j-vanetten replied:
I haven't tested that.  But I know you can physically hold down buttons on the wheel without faults.

xps-genesis replied:
ya, but the steering wheel is already sending a msg continuously and OP is trying to double send the same message when OP wants to send cancel, so the receiving module may fault out if it see discrepancy in the counter at some point. because it is see a counter value from steering button and OP. we may not see any issue if the cancel from OP got accepted quickly, which is what we are mostly seeing on our vehicles.

adeebshihadeh commented on selfdrive/car/chrysler/carstate.py:
```diff
@@ -96,6 +98,7 @@ def get_can_parser(CP):
       ("COUNTER", "EPS_STATUS", -1),
       ("TRACTION_OFF", "TRACTION_BUTTON", 0),
       ("SEATBELT_DRIVER_UNLATCHED", "SEATBELT_STATUS", 0),
+      ("COUNTER", "WHEEL_BUTTONS", 0),
```
add this to the checks

xps-genesis committed:
Update carstate.py

xps-genesis committed:
Update carcontroller.py

j-vanetten commented on selfdrive/car/chrysler/chryslercan.py:
```diff
@@ -48,10 +48,10 @@ def create_lkas_command(packer, apply_steer, moving_fast, frame):
   values = {
     "ACC_CANCEL": cancel,
-    "COUNTER": frame % 10
+    "COUNTER": counter
```
Since we are adding 1 to the `WHEEL_BUTTONS`'s `COUNTER`, we need a mod 16.  This is in case the `COUNTER` is 15 and when we add 1, it should be 0 and not 16 since the `COUNTER` is 0-15

```suggestion
    "COUNTER": counter % 16
```

xps-genesis replied:
since counter itself is a 4 bit signal, it is not necessary to do the % 16, the signal will wrap around when packed on to the CAN bus. 16 will be went as 0 and 17 as 1 and so on.. but either way it is correct.

j-vanetten commented:
kudos to @xps-genesis for making the PR.

xps-genesis committed:
Update selfdrive/car/chrysler/chryslercan.py

since counter itself is a 4 bit signal, it is not necessary to do the % 16, the signal will wrap around when packed on to the can bus. 16 will be went as 0 and 17 as 1 snd so on.. either way it is correct.

Co-authored-by: John Van Etten <4vanetten@gmail.com>
